README.md

```diff
@@ -0,0 +1,237 @@
+
+```java
+public class SimpleTestngTest extends TestcontainersBaseTest {
+
+    private Container alpine = Container.builder()
+            .withImage("alpine:3.5")
+            .build();
+
+    @Test
+    public void doNothing() {
+        alpine.exec("date");
+    }
+}
+```
+
+Note that:
+
+* we use the plain `Container` class as a field in the test class
+* The current (very rough) `TestcontainersBaseTest` class simulates rule-like behaviour: it iterates over fields via reflection and triggers start/stop at appropriate points in the test. This is likely to need a lot of refinement.
+* It's possible that a similar approach might work for JUnit 5 in conjunction with `@ExtendWith`, but this needs proper investigation
+
+
+
+---
+
+## Specialised container classes
+
+Some containers are greatly enhanced by addition of a handful of specialised methods; the database container types are one example.
+
+### End result / usage
+
+```java
+public class SimpleJUnit4Test {
+
+    @Rule
+    public MySqlContainerTestRule mysql = MySqlContainerTestRule.builder()
+            .withImage("mysql:5.5")
+            .withExposedPort(1234)
+            .build();
+
+    @Test
+    public void doSomethingWithMysqlTest() {
+        final String username = mysql.getUsername();
+    }
+```
+
+
+
+We can again accomplish a similar API to Testcontainers 1.x. The example above illustrates a getter for JDBC username on a MySql-specialized class.
+
+
+
+### Implementation
+
+Note that:
+
+* To avoid the mental overhead of generics that current exists in Testcontainers 1.x, here the `MySqlContainerTestRule` class is *not* a subclass of `ContainerTestRule`. Instead, it inherits a default JUnit `apply` method from the `TestRuleImplementation` interface.
+* As with `ContainerTestRule`, the `MySqlContainerTestRule` class _delegates_ to a `Container` instance using the Lombok `@Delegate` annotation.
+* `MySqlContainerTestRule` does some further composition to customize the `Container` instance without resorting to subclassing - it binds a `Plugin` . In this example, the MySql-specific plugin class could do things such as setting environment variables and selecting a readiness check strategy. The next section descibes what a `Plugin` is!
+
+
+
+---
+
+## Plugins
+
+Plugins are intended to allow extra behaviour to be added to `Container` via composition rather than inheritance. This should make it easier to share small, reusable, elements without tying them in to a particular branch of an inheritance hierarchy.
+
+### End result / usage
+
+This is draft and highly subject to change, but the Plugin API could look like the following.
+
+For the developer, adding a plugin to a container would be as simple as calling the `withPlugin` method on the builder, e.g. for a fictitious plugin that adds FakeTime to containers:
+
+```java
+final Container containerWithPlugin = Container.builder()
+  .withImage("alpine:3.5")
+  .withPlugin(new FakeTimePlugin("2010-01-01 00:00:00"))
```

Comment:
maybe `extension` would be a better name? So that we extend container by composing different extensions, but the system itself is not pluggable, at least not with this API

Reply:
I think both terms are fine and apply here, the Wikipedia definition of plugin:
> In computing, a plug-in (or plugin, add-in, addin, add-on, addon, or extension) is a software component that adds a specific feature to an existing computer program.

We add functionality or behavior to the container, but we could also extend the existing feature set ;)

There is also the plugin definition by Martin Fowler in [_Patterns of Enterprise Application Architecture_](https://martinfowler.com/eaaCatalog/plugin.html) which would fit to this implementation I think (basically meaning DI).

Reply:
I don't mind; 'plugin' seemed more natural to me.

One thing which occurred to me is that this isn't essential functionality from day one (it's a new, rather than existing, feature).

Perhaps let's not spend too much time worrying about the details for now, as long as we can agree we'd like this concept to exist in some form 😄

Reply:
Yes, since the fluent builder API would allow us to introduce such functionality without any breaking changes we can simply defer the discussion :slightly_smiling_face: